### ISSUE TITLE
Add option to move the application from internal memory to sdcard 

### ISSUE THREAD

**Issue by ataliba** [GitJournal/GitJournal#99]
Or we have an option for this on the app ? 

Outra possibilidade ao invés de mover os arquivos do git ( repositório ) para o sdcard.

**Comment by TheLastGimbus** [comment]
I can try to do that. I'll PR if I succeed :)

**Comment by vHanda** [comment]
Hey Mateusz

I would love the help. Let me rough out approximately what needs to be done for it -

1. Currently the path is saved in the SharedPreferences from `appstate.dart`. That state will need to be changed and saved(). The next time the app starts, it will be use that path instead.
2. Some settings UI for the user to choose where to save it instead.
3. Someway for this change to affect the currently running app. Either by just restarting the app or some other mechanism.

(3) can be challenging, so maybe you want to tackle it last, and deal with the more straightforwards tasks first. And I can make (3) more of a priority for me, and if I haven't gotten it by the time you have (1), and (2) done, then you can try to do it (if you want to).

Also, I would need you to assign the copyright to me. I'll figure out a simple way for you to do that. I am not sure [if Apple allows or does not allow](https://apple.stackexchange.com/questions/6109/is-it-possible-to-have-gpl-software-in-the-mac-app-store) and therefore would like to error on the side of releasing it as a proprietary app in app store.

**Comment by TheLastGimbus** [comment]
I can't compile app out of the box :/

```
Compiler message:
lib/error_reporting.dart:12:8: Error: Error when reading 'lib/.env.dart': No such file or directory
import 'package:gitjournal/.env.dart';
       ^
lib/screens/purchase_screen.dart:5:8: Error: Error when reading 'lib/.env.dart': No such file or directory
import 'package:gitjournal/.env.dart';
       ^
lib/iap.dart:5:8: Error: Error when reading 'lib/.env.dart': No such file or directory
import 'package:gitjournal/.env.dart';
       ^
lib/error_reporting.dart:17:10: Error: Getter not found: 'environment'.
    dsn: environment['sentry'],
         ^^^^^^^^^^^
lib/screens/purchase_screen.dart:27:27: Error: The getter 'environment' isn't defined for the class '_PurchaseScreenState'.
 - '_PurchaseScreenState' is from 'package:gitjournal/screens/purchase_screen.dart' ('lib/screens/purchase_screen.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'environment'.
    await Purchases.setup(environment['revenueCat']);
                          ^^^^^^^^^^^
lib/iap.dart:16:27: Error: Getter not found: 'environment'.
    await Purchases.setup(environment['revenueCat']);
                          ^^^^^^^^^^^
Target kernel_snapshot failed: Exception: Errors during snapshot creation: null
build failed.

FAILURE: Build failed with an exception.

* Where:
Script '/home/mateusz/dev/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 780

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDevDebug'.
> Process 'command '/home/mateusz/dev/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 20s
Finished with error: Gradle task assembleDebug failed with exit code 1
```

**Comment by vHanda** [comment]
I won't be able to provide a fix till tomorrow. Going to bed right now.

If you want to fix it yourself, generate that missing file by looking at
how the CI does it (since the CI, is green). What I can do tomorrow is try
to add a way so that the file doesn't need to be generated.

**Comment by TheLastGimbus** [comment]
Okay, just needed to run
`flutter/bin/cache/dart-sdk/bin/dart ./scripts/setup_env.dart `
Didn't really touched CI before :P

**Comment by TheLastGimbus** [comment]
Okay, after some coding, I have some proposition:
1. Create new option in settings to select storage location: internal/external. May add more options in future - for example, external but app-specific "Android/data/app.com" that doesn't require storage.
I see that this app is also on iPhone - i have no idea how to deal with storage on iPhone, never had one :/

2. Decide some high-level modifications, in for example `NotesFolderFS` - on how to read/write notes, select folder etc.
Everything will be pretty hard to do cross platform with Android and iOS.
And, you need to remember that Android 11 is coming - and Scoped Storage with it -_-

Honestly, your project is more complicated than I expected, and I'm just starting with Flutter. Gonna make PR that adds storage selection option, but I don't think I can do anything more for now - maybe with some more directions.

**Comment by vHanda** [comment]
This is a great start! GitJournal's code has gotten a bit messy. So don't be too hard on yourself. It's not easy jumping in to a project.

1. About iPhone - I do have one I use for testing, but I have very limited knowledge about the storage capabilities over there. I'll do some research. This can always just be an Android only option as well.

2. I'm a bit confused why `NotesFolderFS` would need to be changed. I assumed that in both internal/external storage, it's just a standard file system. So it's just a matter of telling `NotesFolderFS` which folder to read, and - of course - making sure we have the permissions to read them.

I didn't know about the ScopedStorage in Android 11. I'll read up.

Thanks for the PR!

**Comment by ataliba** [comment]
Hey thanks for this change @TheLastGimbus  :-)
And thanks @vHanda  :-)

**Comment by TheLastGimbus** [comment]
> both internal/external storage, it's just a standard file system

That's just about to change with Android 11. Scoped Storage will make it that you can't read/write outside of what user specified through a file picker. And it will provide some "documents" objects - not standard files :(
And I also had in mind that iOS can have something similar, that's why it would be nice to have some high-lever note interface - like NotesFolderFS - to easly change it later.

**Comment by happycoder97** [comment]
@TheLastGimbus it is not possible to use the result of `ACTION_OPEN_DIRECTORY_TREE` directly from Flutter right? Because they return Uri's instead of paths?

**Comment by TheLastGimbus** [comment]
I didn't respond to this for long, but I don't really know

Although Google said that it listened to community, and offered a "file interfere" between "documents API" - meaning that you can get native `file` object from their cool shiny `document` object, and use that to read/write

**Comment by vHanda** [comment]
So, minor update -

* GitJournal's code base easily supports storing the repo in the directory of your choice.
* However with the underlying framework (Flutter), I haven't found a way to get the SD-Card's location. The Flutter team is aware of the bug - https://github.com/flutter/flutter/issues/40504 - but I don't see any progress being made.
* I've tried other packages - https://github.com/aakashkondhalkar/external_path - but they all only give me the internal storage path. (This is on the emulator)

* At this point I need to learn more about Android Internals and figure out why I can't get the sd card path. Any help would be appreciated.

**Comment by TheLastGimbus** [comment]
Can't you use native folder picker? It should allow user to select wherever they want...

**Comment by vHanda** [comment]
Honestly, I have no idea how to do that with Flutter.

I spent some 20 minutes trying to fix this and I thought I should just
write down my findings, for the next time I decide to investigate it.

I'm sure it's not impossible.

**Comment by itsfarseen** [comment]
Since Android 10, apps can't request write access to a folder in SD Card.  
Instead they will have to ask Storage Access Framework (SAF) to ask user for a directory.  
SAF will show a folder picker and return a path. And our app will be able to do any read/write under that path.  

Currently Flutter standard library doesn't provide a way to access SAF. And as far as I know, there are no 3rd party libraries that do either.

Here's a snippet of code that someone used to invoke SAF folder picker:  
https://github.com/miguelpruivo/flutter_file_picker/issues/721#issuecomment-900100134

They use some kind of channel to pass data to Java/Kotlin part of the app, which has access to SAF.  
The Java/Kotlin code invokes the SAF, gets result and pass it back to flutter.  

The above issue is on `flutter_file_picker` library, which still doesn't support SAF.  
I think it uses the standard library functions to list stuff in a folder that we already have access to.

Ignore any of the comments in that issue mentioning `WRITE_EXTERNAL_STORAGE`. It is a no-op since Android 10.

**Comment by vHanda** [comment]
@itsfarseen : Thank you for all the info.

Minor correction - This is from Android 11, not 10, right?

GitJournal currently uses libgit2 (c code) to perform git operations and doing this via the SAF would be challenging, and it's not something I'm willing to invest the time

I requested the new "manage external storage" permission in Oct, Google denied it, and also kept blocking updates in Gitjournal for 2 months after that for "mysterious" reasons.

I'm planning on improving the settings screen in Android, and tell the user what can and cannot be done. Additionally, all future builds via FDroid will have the extra permission which will allow Gitjournal users to store the git repo wherever they want.

In parallel, I'm also planning on adding "compatibility with other apps" as a core feature and advertise it everywhere, AND monitor the files for changes. With both these changes, maybe Google will grant me the "MANAGE_EXTERNAL_STORAGE" permission, as monitoring files for changes isn't something technically possible via the SAF. (I could be mistaken).

**Comment by itsfarseen** [comment]
> Minor correction - This is from Android 11, not 10, right?

Yes. My bad :)

> GitJournal currently uses libgit2 (c code) to perform git operations and doing this via the SAF would be challenging.

I have no idea whether this would work, but I would try throwing whatever path that SAF returns, to the C code and see if it works. It might work because after all it's a path that SAF returns. If its readable by dart code, it should be readable by the C code as well. Not sure though.

> Additionally, all future builds via FDroid will have the extra permission which will allow Gitjournal users to store the git repo wherever they want.

This is a very good option. Just show it somewhere in the Play Store build, something like "Get the app via FDroid to let you store stuff in arbitrary folders".  

This makes sense, because the reason Google limiting `MANAGE_EXTERNAL_STORAGE` permission is because it's difficult for them to verify that the app doesn't do malicious stuff like ftp all the user's private files to a remote server.  
Since we are open source, and FDroid guarantees that an APK corresponds to a particular source version, we can be sure that the APK doesn't do such stuff by simply reading the code. 

> as monitoring files for changes isn't something technically possible via the SAF

Yeah I think so too.

> maybe Google will grant me the "MANAGE_EXTERNAL_STORAGE" permission

I'm not very hopeful about this though.

**Comment by TheLastGimbus** [comment]
Guys, guys

Google made a big poo poo, but TL;DR: 
- on Android <= 10 - use `android:requestLegacyExternalStorage="true"` flag, and use native file paths
- on Android >= 11, you need to use SAF, but it can give you access to native path. It will fully work with all C++-etc libraries :rainbow: You can watch this talk, aspecially `15:02`: https://youtu.be/UnJ3amzJM94?t=902

There probably already should be some SAF plugin somewhere... yes, you need to access it with Java (plugin), but once you get stuff from the user, you can get normal `/storage/emulated/0/something` path and be happy :+1:
